src/lib.rs

```diff
@@ -168,6 +168,53 @@ pub trait StructOfArray {
     type Type;
 }
 
+
+mod private_soa_indexs {
+    // From [`std::slice::SliceIndex`](https://doc.rust-lang.org/std/slice/trait.SliceIndex.html) code.
+    // Limits the types that may implement the SoA index traits.
+    // It's also helpful to have the exaustive list of all accepted types.
+
+    use ::std::ops;
+
+    pub trait Sealed {}
+
+    impl Sealed for usize {}                        // [a]
+    impl Sealed for ops::Range<usize> {}            // [a..b]
+    impl Sealed for ops::RangeTo<usize> {}          // [..b]
+    impl Sealed for ops::RangeFrom<usize> {}        // [a..]
+    impl Sealed for ops::RangeFull {}               // [..]
+    impl Sealed for ops::RangeInclusive<usize> {}   // [a..=b]
+    impl Sealed for ops::RangeToInclusive<usize> {} // [..=b]
+}
+
+/// Helper trait used for indexing operations.
+/// Inspired by [`std::slice::SliceIndex`](https://doc.rust-lang.org/std/slice/trait.SliceIndex.html).
+pub trait SoaIndex<T>: private_soa_indexs::Sealed {
```

Comment:
naming: how about `SoAIndex` and `SoAIndexMut`?

Reply:
Sure